File: leetcode/sort-integers-by-the-number-of-1-bits.py
class Solution(object):
    def sortByBits(self, arr):

        arr.sort(key = lambda x :(bin(x).count("1"), x))

        return arr


if __name__ == "__main__":
    solver = Solution()
    
    def run_test(case_num, arr, expected):
        result = solver.sortByBits(arr[:])
        status = "PASS" if result == expected else "FAIL"
        
        print(f"Test Case {case_num}:")
        print(f"  Input:    {arr}")
        print(f"  Output:   {result}")
        print(f"  Expected: {expected}")
        print(f"  Status:   {status}\n")

    # Example 1
    # 0 -> 0 bits
    # 1, 2, 4, 8 -> 1 bit
    # 3, 5, 6 -> 2 bits
    # 7 -> 3 bits
    run_test(1, [0,1,2,3,4,5,6,7,8], [0,1,2,4,8,3,5,6,7])

    # Example 2
    # Powers of 2 all have exactly 1 bit. 
    # They should just be sorted by their actual values.
    run_test(2, [1024,512,256,128,64,32,16,8,4,2,1], [1,2,4,8,16,32,64,128,256,512,1024])

    # Custom Case: Duplicate values
    run_test(3, [10, 10, 3, 3], [3, 3, 10, 10])
    
    # Custom Case: Same bit counts, distinct large numbers
    # 15 (1111) -> 4 bits, 23 (10111) -> 4 bits
    run_test(4, [23, 15], [15, 23])

File: golang/sort-integers-by-the-number-of-1-bits/sort-integers-by-the-number-of-1-bits.go
package main

import (
	"fmt"
	"math/bits"
	"reflect"
	"sort"
)

func sortByBits(arr []int) []int {
	sort.Slice(arr, func(i, j int) bool {
		bitsI := bits.OnesCount(uint(arr[i]))
		bitsJ := bits.OnesCount(uint(arr[j]))

		if bitsI == bitsJ {
			return arr[i] < arr[j]
		}

		return  bitsI < bitsJ
	})

	return arr
}

func runTest(caseNum int, arr []int, expected []int) {
	input := make([]int, len(arr))
	copy(input, arr)

	result := sortByBits(input)

	status := "FAIL"
	if reflect.DeepEqual(result, expected) {
		status = "PASS"
	}

	fmt.Printf("Test Case %d:\n", caseNum)
	fmt.Printf("  Input:    %v\n", arr)
	fmt.Printf("  Output:   %v\n", result)
	fmt.Printf("  Expected: %v\n", expected)
	fmt.Printf("  Status:   %s\n\n", status)
}

func main() {
	runTest(1,
		[]int{0, 1, 2, 3, 4, 5, 6, 7, 8},
		[]int{0, 1, 2, 4, 8, 3, 5, 6, 7},
	)

	runTest(2,
		[]int{1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1},
		[]int{1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024},
	)

	runTest(3,
		[]int{10, 10, 3, 3},
		[]int{3, 3, 10, 10},
	)

	runTest(4,
		[]int{23, 15},
		[]int{15, 23},
	)
}
